### PR TITLE
Add support to enable/disable strict dependency checking

### PR DESCRIPTION
**Pull request by arnaudsjs**
# Description

This PR add support to enable/disable strict dependency checking in the compiler and in the module tools. I will create a similar PR on ISO5 when this one gets accepted.

Part of #4409

# Self Check:

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/module.py [hunk 1]
```diff
@@ -2009,14 +2024,21 @@ def verify_python_requires(self) -> None:
...
+            constraints: List[Requirement] = [Requirement.parse(item) for item in self.collect_python_requirements()]
+            try:
+                env.ActiveEnv.check(strict_scope=re.compile(f"{ModuleV2.PKG_NAME_PREFIX}.*"), constraints=constraints)
+            except env.ConflictingRequirements as e:
+                message: str = "Module dependency resolution conflict: a module dependency constraint \
+    was violated by another module. This most likely indicates an incompatibility between \
+    two or more of the installed modules."
+                raise env.ConflictingRequirements(message, e.conflicts)
+        else:
+            if not env.ActiveEnv.check_legacy(in_scope=re.compile(f"{ModuleV2.PKG_NAME_PREFIX}.*")):
```
Comment (sanderr) [inline src/inmanta/module.py:2037]: Nice catch, I just realized that I'd forgotten about this in the ticket.

File: changelogs/unreleased/4409-add-support-disable-strict-checking.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+description: Add support to enable/disable strict dependency checking in the compiler and in the module tools using the --strict-deps-check and --no-strict-deps-check options.
+issue-nr: 4516
+change-type: minor
+destination-branches: [master]
+sections:
+  minor-improvement: "{{description}}"
```
Comment (sanderr) [inline changelogs/unreleased/4409-add-support-disable-strict-checking.yml:7]: Maybe add an upgrade note as well?

File: src/inmanta/env.py [hunk 3]
```diff
@@ -62,20 +62,36 @@ class PackageNotFound(Exception):
     pass
 
 
+@dataclass(eq=True, order=True, frozen=True)
+class VersionConflict:
```
Comment (FloLey) [inline src/inmanta/env.py:66]: This is so much better than before :p Also I learned about dataclass frozen :)

File: tests/test_app_cli.py [hunk 4]
```diff
@@ -66,10 +66,12 @@ async def install_project(python_env: env.PythonEnvironment, project_dir: py.pat
...
     )
     try:
-        (stdout, stderr) = await asyncio.wait_for(process.communicate(), timeout=30)
+        await asyncio.wait_for(process.communicate(), timeout=30)
     except asyncio.TimeoutError as e:
         process.kill()
-        await process.communicate()
+        (stdout, stderr) = await process.communicate()
+        print(stdout.decode())
+        print(stderr.decode())
```
Comment (FloLey) [inline tests/test_app_cli.py:74]: I think this can be removed (at least once)

**Review by FloLey — APPROVED** [review]

File: tests/test_app_cli.py (hunk 4 above)
Comment (arnaudsjs) [reply tests/test_app_cli.py:74]: What do you mean?

**Pull request closed without merging**